File: Part1_강의자료/Ch04_전체탐색_시뮬레이션/문제별코드/2840_행운의바퀴/2840.py
N, K = map(int, input().split())
cur_idx = 0
ans = ['?' for i in range(N)]
check = [0 for i in range(26)]

for i in range(K):
    S, C = input().split()
    S = int(S)
    next_idx = ((cur_idx - S) % N + N) % N
    if ans[next_idx] != '?' and ans[next_idx] != C:
        print("!")
        break
    elif ans[next_idx] == '?' and C in ans:
        print("!")
        break
    ans[next_idx] = C
    cur_idx = next_idx

    if i == K - 1:
        ans = ans[cur_idx:] + ans[:cur_idx]
        print("".join(ans))

File: Part1_강의자료/Ch04_전체탐색_시뮬레이션/문제별코드/2840_행운의바퀴/2840.cpp
#include <stdio.h>
#include <string.h>

int N, K, S;
char C;
char ans[25];
int check[26];

int main()
{
    scanf("%d %d", &N, &K);

    int cur_idx = 0;
    memset(ans, '?', sizeof(ans));
    while (K--)
    {
        scanf("%d %c", &S, &C);
        int next_idx = ((cur_idx - S) % N + N) % N;
        if (ans[next_idx] == '?')
            ans[next_idx] = C;
        else if (ans[next_idx] != C)
        {
            printf("!\n");
            return 0;
        }
        cur_idx = next_idx;
    }

    for (int i = 0; i < N; i++)
    {
        if (ans[i] == '?')
            continue;
        if (check[ans[i] - 'A'])
        {
            printf("!\n");
            return 0;
        }
        check[ans[i] - 'A'] = 1;
    }

    for (int i = 0; i < N; i++)
        printf("%c", ans[(cur_idx + i) % N]);
    printf("\n");
    return 0;
}
